### PR TITLE
Feature/sli blk traversal

### PR DESCRIPTION
**Pull request by Elementarereigniss**
# Description

Create a 3D slice traversal, based on the described sli_blk traversal in "Algorithmic and Implementational Optimizations of Molecular Dynamics Simulations for Process Engineering" chapter 8.4, but adapted for 3 dimensional slicing.

# How Has This Been Tested?

- [x] UnitTests
- [x] md-flexible execution with specific yaml-file

### CODE REVIEW

File: src/autopas/containers/linkedCells/traversals/SlicedBlkTraversal.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * This class provides the multiple dimension sliced traversal.
+ *
+ * @tparam ParticleCell
+ * @tparam PairwiseFunctor
+ * @tparam dataLayout
+ * @tparam useNewton3
+ */
+template<class ParticleCell, class PairwiseFunctor, DataLayoutOption::Value dataLayout, bool useNewton3>
+class SlicedBlkTraversal : public SlicedBlkBasedTraversal<ParticleCell, PairwiseFunctor, dataLayout, useNewton3>,
```
Comment (FG-TUM) [inline src/autopas/containers/linkedCells/traversals/SlicedBlkTraversal.h:32]: What's the reason for splitting this traversal in a SlicedBlkBases and the real one? Do you plan to implement further versions that work on the same base? If no merge the two files.

**Review by FG-TUM — COMMENTED** [review]
- [x] needs clang-format
- [x] There seem to be many notes in the comments. Try to clean them up such that only comments remain that actually explain what you do.

**Comment by FG-TUM** [comment]
Before I review this please resolve the merge conflicts.

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+    }
+    return str;
+  }
+  std::string debugHelperFunctionIntArrayToString(std::array<double, 3> intArray) {
+    std::string str;
+    for (auto i : intArray) {
+      str += std::to_string(i) + " ";
+    }
+    return str;
+  }
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:298]: remove those. If they are really needed for some logging an equivalent function already exists in ArrayUtils

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+      for (unsigned long y = 0; y < 2; ++y) {
+        for (unsigned long z = 0; z < 2; ++z) {
+          LocksThatCanBeSet[no_locksSet] =
+              locks[uniqueLockCalculation(blockNumber, {x + order[0], y + order[1], z + order[2]})].testlock();
+          //          AutoPasLog(
+          //              debug,
+          //              "TRYING LOCKING " + std::to_string(blockNumber) + " Lock : " +
+          //                  std::to_string(uniqueLockCalculation(blockNumber, {x + order[0], y + order[1], z +
+          //                  order[2]})) + " id: " + debugHelperFunctionIntArrayToString2({x + order[0], y + order[1],
+          //                  z + order[2]}));
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:325]: remove any commented-out code

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+          no_locksSet++;
+        }
+      }
+    }
+
+    if (LocksThatCanBeSet[0] && LocksThatCanBeSet[1] && LocksThatCanBeSet[2] && LocksThatCanBeSet[3] &&
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:331]: to be consistent with the rest of autopas replace all `&&` with `and` and all `||` with `or`

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+        for (int i = 0; i < 2; ++i) {
+          for (int j = 0; j < 2; ++j) {
+            if (axis == 0) {
+              AutoPasLog(debug, "UNLOCKING: " + std::to_string(blockNumber) + " Lock : " +
+                                    std::to_string(uniqueLockCalculation(
+                                        blockNumber, {orderCurrent[0] + diffOrderCurrent, i + orderCurrent[1],
+                                                      j + orderCurrent[2]})) +
+                                    " id: " +
+                                    debugHelperFunctionIntArrayToString2({orderCurrent[0] + diffOrderCurrent,
+                                                                          i + orderCurrent[1], j + orderCurrent[2]}));
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:410]: remove any debug output that is not useful for any plotting.

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 6]
```diff
@@ -0,0 +1,953 @@
...
+          blocks[cellBlockIterator][1][1] = y + _cellBlockDimensions[1][yit];
+        } else {
+          blocks[cellBlockIterator][1][1] = y + _cellBlockDimensions[1][yit] - _overlapAxis[1];
+        }
+
+        if (z + _cellBlockDimensions[2][zit] < _dims[2] - _overlapAxis[2]) {
+          blocks[cellBlockIterator][1][2] = z + _cellBlockDimensions[2][zit];
+        } else {
+          blocks[cellBlockIterator][1][2] = z + _cellBlockDimensions[2][zit] - _overlapAxis[2];
+        }
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:749]: if you combine `x`, `y`, `z` into an array you can reduce this wall of code to a loop (might be true at other places too)

File: src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h [hunk 7]
```diff
@@ -0,0 +1,953 @@
...
+    }
+
+    // Iterate over own sub-blocks (excluding all with 2)
+    while (!traversalOrderQueue->empty()) {
+      // todo: maybe make an indicator if a broken traversalOrder was used e.g. >10.000 Iterations per thread
+      //  or tell scheduler to give thread less priority after x iterations of this
+      int traversalOrder = traversalOrderQueue->front();
+      traversalOrderQueue->pop();
+      auto &order = _cellBlockTraverseOrderByLocks[traversalOrder];
+      if (!locking2x2x2sub_blocks(m, order)) {
```
Comment (FG-TUM) [inline src/autopas/containers/cellPairTraversals/SlicedBlkBasedTraversal.h:913]:
```suggestion
      if (not locking2x2x2sub_blocks(m, order)) {
```

**Review by FG-TUM — CHANGES_REQUESTED** [review]
Please do some cleanup before I do a proper review:
- [x] remove the input files
- [x] remove debug output
- [x] condense repetitive code into loops or (lambda-)functions
- [x] remove unused code
- [ ] resolve merge conflicts from master

**Comment by SteffenSeckler** [comment]
not in a mergeable state, requires reimplementation.

**Pull request closed without merging**